glean-core/rlb/src/private/ping.rs

```diff
@@ -36,15 +45,43 @@ impl PingType {
             reason_codes,
         );
 
-        let me = Self { name, ping_type };
+        let me = Self {
+            name,
+            ping_type,
+            test_callback: Arc::new(Default::default()),
+        };
         crate::register_ping_type(&me);
         me
     }
+
+    /// **Test-only API**
+    ///
+    /// Attach a callback to be called right before a new ping is submitted.
+    /// The provided function is called exactly once before submitting a ping.
+    ///
+    /// Note: The callback will be called on any call to submit.
+    /// A ping might not be sent afterwards, e.g. if the ping is otherwise empty (and
+    /// `send_if_empty` is `false`).
+    ///
+    /// TODO: Ensure this is only usable in test mode.
+    pub fn test_before_next_submit(&self, cb: impl FnOnce(Option<&str>) + Send + 'static) {
+        let mut test_callback = self.test_callback.lock().unwrap();
+
+        let cb = Box::new(cb);
+        *test_callback = Some(cb);
+    }
 }
 
 #[inherent(pub)]
 impl glean_core::traits::Ping for PingType {
     fn submit(&self, reason: Option<&str>) {
+        // TODO: Ensure this is only usable in test mode.
```

Comment:
Again: We have a test mode?!

Reply:
No, we don't in the RLB.
We do in other bindings and are thus able to assert when using `testBeforeNextSubmit`. No such protection is given in the RLB.